File: android/app/src/main/java/com/dungeonboss/game/phases/BuildPhase.kt
package com.dungeonboss.game.phases

import com.dungeonboss.game.Game
import com.dungeonboss.game.Player
import com.dungeonboss.model.BuildCard
import com.dungeonboss.model.Room

/**
 * Build phase. A player MAY take one build action: place a room into any of the
 * 5 slots (an empty slot is filled, an occupied slot is replaced), or spend a
 * basic/advanced room card to upgrade a placed room (granting its bait icons and
 * a room level). This phase exposes the step the decision loop drives.
 * Orchestration only.
 *
 * [place] target encoding:
 *   - placing a room        -> the slot index (Int 0..4)
 *   - upgrading with a room  -> the string "upgrade:<slot>" (the Room is spent)
 */
object BuildPhase {

    /**
     * Apply a player's build choice. Returns the cards this action pushed to the
     * room deck (a replaced room, or the spent upgrade card), so the caller can
     * reclaim them when undoing the placement.
     */
    fun place(game: Game, player: Player, cardId: String?, target: Any?): List<BuildCard> {
        if (cardId == null) return emptyList()

        val card = player.takeRoomFromHand(cardId)
            ?: throw IllegalArgumentException("card not in hand: $cardId")
        val dungeon = player.dungeon!!
        val discarded = mutableListOf<BuildCard>()

        when {
            card is Room && isUpgradeWith(target) -> {
                // Spend a basic/advanced room card to upgrade a placed room: grants
                // its bait icons and a room level. The spent card is discarded.
                dungeon.upgradeRoomWith(slotOf(target), card)
                game.roomDeck.discard(card)
                discarded.add(card)
            }
            card is Room && card.advanced -> placeAdvanced(game, player, card, slotOf(target), discarded)
            card is Room -> placeBasic(game, dungeon, card, slotOf(target), discarded)
        }
        return discarded
    }

    private fun placeBasic(
        game: Game,
        dungeon: com.dungeonboss.game.Dungeon,
        card: Room,
        slot: Int,
        discarded: MutableList<BuildCard>
    ) {
        val old = dungeon.placeRoom(slot, card)
        if (old != null) {
            game.roomDeck.discard(old.baseRoom); discarded.add(old.baseRoom) // replaced room returns to deck
        }
    }

    /**
     * An advanced room may fill an empty slot freely; replacing an occupied slot
     * requires that room to share at least one of the advanced room's bait icons.
     */
    private fun placeAdvanced(game: Game, player: Player, card: Room, slot: Int, discarded: MutableList<BuildCard>) {
        val dungeon = player.dungeon!!
        val existing = dungeon.slots[slot]
        if (existing != null) {
            require(existing.bait.shares(card.bait)) {
                "advanced room needs a room sharing at least one bait icon"
            }
        }
        placeBasic(game, dungeon, card, slot, discarded)
    }

    private fun isUpgradeWith(target: Any?): Boolean =
        target is String && target.startsWith("upgrade:")

    private fun slotOf(target: Any?): Int = when (target) {
        is Int -> target
        else -> target.toString().substringAfter(':', target.toString()).trim().toInt()
    }
}

File: webapp/lib/phases/build_phase.rb
# frozen_string_literal: true

require_relative "../upgrade"

# Build phase. For each player:
#   - draw two room cards into hand
#   - the player CHOOSES one room from hand to discard
#   - the player CHOOSES one room from hand to place, or to build nothing
#
# A room may be placed as a new entrance (to the left) when the dungeon is not
# full, or placed on top of an existing room to replace it (the replaced room is
# discarded). A dungeon holds at most Dungeon::MAX_ROOMS rooms.
#
# This phase exposes the steps the game's decision loop drives; it does not pick
# for the player. Orchestration only.
class BuildPhase
  DRAW_COUNT = 2

  # Draw two rooms into each player's hand, up to the hand cap; any room that
  # does not fit is discarded back to the deck.
  def self.draw_for_each(game)
    game.living_players.each do |player|
      game.room_deck.draw_many(DRAW_COUNT).each do |room|
        game.room_deck.discard(room) unless player.add_room_to_hand(room)
      end
    end
  end

  # Discard a chosen room from the player's hand (mandatory each Build phase).
  # Returns the discarded card so the caller can offer an undo.
  def self.discard(game, player, room_id)
    room = player.take_room_from_hand(room_id)
    raise ArgumentError, "room not in hand: #{room_id.inspect}" unless room

    game.room_deck.discard(room)
    room
  end

  # Apply a player's build choice.
  #   card_id: the card from hand to play, or nil to build nothing
  #   target:  basic room -> "new" (or nil) to add an entrance, or a room index
  #            to replace; upgrade or advanced room -> the room index it targets.
  def self.place(game, player, card_id, target = nil)
    return if card_id.nil?

    card = player.take_room_from_hand(card_id)
    raise ArgumentError, "card not in hand: #{card_id.inspect}" unless card

    if card.is_a?(Upgrade)
      replaced = player.dungeon.apply_upgrade(Integer(target), card)
      game.room_deck.discard(replaced) if replaced # an upgrade it displaced
    elsif card.advanced?
      place_advanced(game, player, card, Integer(target))
    elsif target.nil? || target.to_s == "new"
      player.dungeon.add_room_to_left(card)
    else
      discard_replaced(game, player.dungeon.replace_room(Integer(target), card))
    end
  end

  # An advanced room may replace any room that shares at least one of its bait
  # icons.
  def self.place_advanced(game, player, card, index)
    placed = player.dungeon.rooms.fetch(index)
    unless placed.bait.shares?(card.bait)
      raise ArgumentError, "advanced room needs a room sharing at least one bait icon"
    end

    discard_replaced(game, player.dungeon.replace_room(index, card))
  end

  # Return a replaced PlacedRoom's base room (and any upgrade) to the deck.
  def self.discard_replaced(game, old)
    game.room_deck.discard(old.base_room)
    game.room_deck.discard(old.upgrade) if old.upgrade
  end
end
